rust/src/dns/dns.rs

```diff
@@ -659,23 +659,50 @@ impl DNSState {
     }
 }
 
+fn probe_header_validity(header : DNSHeader, rlen: usize) -> (bool, bool, bool) {
+    let opcode = ((header.flags >> 11) & 0xf) as u8;
+    if opcode >= 7 {
+        //unassigned opcode
+        return (false, false, false);
+    }
+    if header.additional_rr as usize
+        + header.answer_rr as usize
+        + header.authority_rr as usize
+        + header.questions as usize
+        > rlen
+    {
+        //not enough data for additional records
+        return (false, false, false);
+    }
```

Comment:
As these are counts not sizes, its not accurate, but also won't false positive.  Should the size of the header be calculated in as well?  If sizeof(header) + addtional_rr +answer_rr > rlen...  Might be a bit more accurate.  I think we could also assume a minumum of 2 bytes per expected record being the CLASS is a u16.

Reply:
Indeed, these are counts, not sizes.
And I do not count the header size.
But as each of these has at least one byte, we can do this test to exclude some cases
 
So, should I include the header size in this sum ?

Reply:
> I think so. Just to confirm, rlen is the size of the record. In TCP this would be the first 2 bytes, in UDP its the payload size.

Exactly, you can check the callers do it right

> So a test to make sure its greater than or equal to:
> 
> ```
> sizeof(header) + (additional_rr * 2) + (answer_rr * 2) + (authority_rr * 2) + (questions * 2)
> ```
> 
> should be safe and accurate, right?

Ok, will do this